build.py

```diff
@@ -0,0 +1,188 @@
+                zip_deploy = ZipFile(export_directory / EXTENSION_ZIP_NAME, 'w', ZIP_DEFLATED, compresslevel=9)
+            else:
+                zip_deploy = contextlib.nullcontext()
+
+            with zip_deploy as zip:
+
+                printe('-' * 80)
+                printe('packaging version', version)
+
+                with BASE_MANIFEST.open('r') as f:
+                    extension_manifest = json.load(f)
+
+                version_files_to_include = filter(lambda f: f.is_file(), version.iterdir())
+
+                for file_to_include in itertools.chain(files_to_include, version_files_to_include):
+                    exported_file = export_directory / file_to_include
+                    exported_file.parent.mkdir(parents=True, exist_ok=True)
+                    try:
+                        # hardlinks take a fraction of time related to copy&paste
+                        # Unfortunately, some platforms don't implement or allow them for regular users (E.g. windows)
+                        file_to_include.link_to(exported_file)
+                        # Don't show the user a hard link is tried when the feature is not implemented
+                        printe('debug: hardlinked', file_to_include)
+                    except NotImplementedError:  # Not implemented in windows
+                        printe('debug: copy', file_to_include)
+                        shutil.copy(file_to_include, exported_file)
+                    except Exception:
+                        printe('debug: hardlink', file_to_include)  # Pretends the hardlink was tried but an error happened
+                        raise
+
+                    if zip:
+                        if file_to_include.suffix == '.js':
+                            printe('release: uglify js', file_to_include)
+                            zip_content = subprocess.check_output([
+                                'uglifyjs', '--compress', '--mangle', '--no-annotations', '-c', 'drop_console',
+                                '--', os.fspath(file_to_include.absolute())
+                            ])
+                        else:
+                            printe('release: include file', file_to_include)
+                            zip_content = file_to_include.read_bytes()
+
+                        zip.writestr(os.fspath(file_to_include), zip_content)
+
+                version_manifest = version / 'manifest.json'
+
+                printe('debug: manifest', version_manifest)
+
+                with version_manifest.open('r') as f:
+                    specific_manifest = json.load(f)
+
+                specific_manifest['permissions'].extend(extension_manifest['permissions'])
+                extension_manifest.update(specific_manifest)
+
+                with (export_directory / 'manifest.json').open('w') as f:
+                    json.dump(extension_manifest, f, indent=4)
```

Comment:
personal preference again, also added bonus of not clearing the entire file if the json dump fails for some reason
```suggestion
                manifest_content = json.dumps(extension_manifest, indent=4)
                (export_directory / 'manifest.json').write_text(manifest_content)
```

Reply:
In my opinion this one is arguable... But it is rather valid. I'll think about it tomorrow.